### ISSUE TITLE
Wrong URL when using CloudFront in front of API Gateway

### ISSUE THREAD

**Issue by mnapoli** [mnapoli/bref#67]
Just reporting a bug I'm working on right now: I have trouble getting the correct URL called in the browser when using CloudFront in front of API Gateway.

```
CloudFront -> API Gateway -> AWS Lambda
```

CloudFront is useful to be able to serve both assets from S3 and dynamic content from Lamdba in the same website.

However the URL detected by Bref in the lambda is not the one actually loaded in Cloudfront because Cloudfront URLs do not contain the stage prefix (e.g. `/dev`).

I have asked a question on stackowerflow: https://stackoverflow.com/questions/52582792/how-to-get-the-original-url-called-in-aws-lambda Here is a description of the problem:

I have a lambda setup with the proxy integration in API Gateway. I can call my lambda with a URL like https://7kbw9fcfa4.execute-api.us-east-1.amazonaws.com/dev/foo The path here is `/dev/foo`.

However if I set up CloudFront in front of that, the URL becomes https://a45ex7tnds5r5o.cloudfront.net/foo and the path is `/foo`.

In both cases I have the same keys in the `event` variable:

- `path` contains `/foo`
- `requestContext.path` contains `/dev/foo`

So I don't know yet how to reliably retrieve the correct URL of the original request.

**Comment by mnapoli** [comment]
This is actually the same issue with a custom domain name I believe.

One thing we could do is always drop the stage prefix (`/dev`) from the URL. That means it's easier to go from local development (e.g. Symfony routes) to online on API Gateway to production with a custom domain/Cloudfront: we use always the same routes.

But that also means that we are "hiding" the real URLs when on API Gateway. And if we use the router to generate URLs (e.g. links to other pages) then we generate wrong links… That sucks.

**Comment by nealio82** [comment]
@eddmann knows AWS architecture pretty well, any thoughts? Can the URI be mapped from one to the other programattically somehow on the AWS side?

**Comment by mnapoli** [comment]
@eddmann yes, but the problem is when using CloudFront like this:

```
CloudFront -> API Gateway -> AWS Lambda
```

**Comment by eddmann** [comment]
@mnapoli Are you using CloudFront directly or via API Gateway?

**Comment by mnapoli** [comment]
I am using it in front of API Gateway.

I use it to serve both assets from S3 and dynamic content from Lamdba in the same website/domain name.

I had an interesting answer on StackOverflow:

> You could configure your CloudFront distribution to use a custom header that would be passed to the Lambda function. Then you could determine whether to use path or requestContext.path based on if that CloudFront-only header is present.

I guess that would work, but that's definitely not great to ask Bref users to do that for their projects…

**Comment by wysow** [comment]
Sounds like a good "hack" which could be at least partially automated, and if we add some information in bref readme that should be fine for now, no?

**Comment by eddmann** [comment]
Ahh right. Yes you could add an additional header with CloudFront requests. Another option is to have two separate CloudFront distros. one which API Gateway manages (handling the desired prefix etc). and another which is used for static assets. Doing it this way however would require two separate domains which I'm assuming you are trying to get away from?

Another option that I haven't tried and am unsure on (due to Lambda sizes) is to store all assets within the Lambda along side the code. In this case you have a dedicated handler that gets run when static assets are requested. Performance for this would not be ideal on initial request but as you can tweak the Cache controls and API Gateway is going through CloudFront you will benefit from there caching platform.

**Comment by mnapoli** [comment]
> Doing it this way however would require two separate domains which I'm assuming you are trying to get away from?

Yes, I'm trying to make it possible to do what most PHP applications do, i.e. serve both PHP and assets under the same domain.

> store all assets within the Lambda along side the code. In this case you have a dedicated handler that gets run when static assets are requested.

I wondered about this too. The first hit would be slow and that seems overkill to serve assets using PHP. And as you mention, it's too bad to have to upload the assets to the lambda, that doesn't seem ideal 🤔 	But maybe for small application it could work… Anyway if I can make it work with CloudFront that's better IMO.

**Comment by wysow** [comment]
Just my 2 cents, lambda zip file is limited to 50mb so all assets should be optimized, and they should be served by the node handler without ever hit the php handler, what do you think?

**Comment by nealio82** [comment]
> Just my 2 cents, lambda zip file is limited to 50mb so all assets should be optimized, and they should be served by the node handler without ever hit the php handler, what do you think?

Hmm, wouldn't that cause a problem when the routing / management for assets is handled by the PHP framework?

**Comment by mnapoli** [comment]
Oh that’s definitely simpler to serve assets using Node! Maybe there could be an optional config to let Bref’s JS handler serve assets, and if you want to handle those manually with Cloudfront or PHP then it’s up to you...

**Comment by nealio82** [comment]
Sure - maybe some configurable toggle to serve assets using node / PHP?

**Comment by mnapoli** [comment]
Should be fixed in 0.3.
